### DISCUSSION TITLE
ROS Foxy - InvalidNode - USB Webcam Feed

### DISCUSSION THREAD

**Discussion by PhilShaw852** [OpenVSLAM-Community/openvslam_ros#154] [Q&A] [answered]
Sorry, back again. Thanks for the rapid answer to my last question.

I start:

ros2 run image_tools cam2image

No issue, then I start:

ros2 run image_transport republish \
    raw in:=image raw out:=/camera/image_raw

Again, no issue, then I start:
ros2 run stella_vslam_ros run_slam \
    -v /home/ps/ros2_ws/src/stella_vslam_ros/vocab/orb_vocab.fbow \
    -c /home/ps/ros2_ws/src/stella_vslam_ros/vocab/config.yaml \
    --map-db-out /home/ps/ros2_ws/src/stella_vslam_ros/map/map.msg \
    --ros-args -p publish_tf:=false

And I get:

[I] config file loaded: /home/ps/ros2_ws/src/stella_vslam_ros/vocab/config.yaml
[2023-08-30 07:55:41.815] [I] 
original version of OpenVSLAM,
Copyright (C) 2019,
National Institute of Advanced Industrial Science and Technology (AIST)
All rights reserved.
stella_vslam (the changes after forking from OpenVSLAM),
Copyright (C) 2022, stella-cv, All rights reserved.

This is free software,
and you are welcome to redistribute it under certain conditions.
See the LICENSE file.

Camera.fx: 517.306408
Camera.fy: 516.469215
Camera.cx: 318.643040
Camera.cy: 255.313989
Camera.k1: 0.262383
Camera.k2: -0.953104
Camera.p1: -0.005358
Camera.p2: 0.002628
Camera.k3: 1.163314
Camera.fps: 30.0
Camera.RGB: 1
ORBextractor.nFeatures: 1000
ORBextractor.scaleFactor: 1.2
ORBextractor.nLevels: 8
ORBextractor.iniThFAST: 20
ORBextractor.minThFAST: 7
Viewer.KeyFrameSize: 0.04
Viewer.KeyFrameLineWidth: 1
Viewer.GraphLineWidth: 0.9
Viewer.PointSize: 0.025
Viewer.CameraSize: 0.08
Viewer.CameraLineWidth: 3
Viewer.ViewpointX: 0
Viewer.ViewpointY: -0.7
Viewer.ViewpointZ: -1.8
Viewer.ViewpointF: 500

[2023-08-30 07:55:41.815] [I] loading ORB vocabulary: /home/ps/ros2_ws/src/stella_vslam_ros/vocab/orb_vocab.fbow
terminate called after throwing an instance of 'YAML::InvalidNode'
  what():  invalid node; this may result from using a map iterator as a sequence iterator, or vice-versa

The vocab file is from[ ](https://github.com/stella-cv/FBoW_orb_vocab/raw/main/orb_vocab.fbow).

Or could the config.yaml be the issue, the one I included is temporary while I figure out how to calibrate.

**Comment by ymd-stella** [comment, marked as answer]
The format of the configuration file is incorrect, see https://github.com/stella-cv/stella_vslam/tree/main/example.

**Reply by PhilShaw852** [reply]
Thanks, I updated the file and I now get a little further.

Now I get:

[2023-09-01 00:02:20.197] [I] config file loaded: /home/ps/ros2_ws/src/stella_vslam_ros/vocab/config.yaml
[2023-09-01 00:02:20.197] [I] 
original version of OpenVSLAM,
Copyright (C) 2019,
National Institute of Advanced Industrial Science and Technology (AIST)
All rights reserved.
stella_vslam (the changes after forking from OpenVSLAM),
Copyright (C) 2022, stella-cv, All rights reserved.

This is free software,
and you are welcome to redistribute it under certain conditions.
See the LICENSE file.

Camera:
  name: Tello Cam
  setup: monocular
  model: perspective
  fx: 517.306408
  fy: 516.469215
  cx: 318.643040
  cy: 255.313989
  k1: 0.262383
  k2: -0.953104
  p1: -0.005358
  p2: 0.002628
  k3: 1.163314
  fps: 30.0
  cols: 960
  rows: 720
  color_order: BGR
Preprocessing:
  min_size: 1000
Feature:
  name: default ORB feature extraction setting
  scale_factor: 1.2
  num_levels: 8
  ini_fast_threshold: 20
  min_fast_threshold: 7
Mapping:
  backend: g2o
  baseline_dist_thr_ratio: 0.02
  redundant_obs_ratio_thr: 0.9
  num_covisibilities_for_landmark_generation: 20
  num_covisibilities_for_landmark_fusion: 20
  erase_temporal_keyframes: false
  num_temporal_keyframes: 15
Tracking:
  backend: g2o
  enable_temporal_keyframe_only_tracking: false
KeyframeInserter:
  wait_for_local_bundle_adjustment: false
Relocalizer:
  search_neighbor: true
LoopDetector:
  backend: g2o
System:
  map_format: msgpack
PangolinViewer:
  keyframe_size: 0.04
  keyframe_line_width: 1
  graph_line_width: 0.9
  point_size: 0.025
  camera_size: 0.08
  camera_line_width: 3
  viewpoint_x: 0
  viewpoint_y: -0.7
  viewpoint_z: -1.8
  viewpoint_f: 500

[2023-09-01 00:02:20.197] [I] loading ORB vocabulary: /home/ps/ros2_ws/src/stella_vslam_ros/vocab/orb_vocab.fbow
[2023-09-01 00:02:20.214] [I] load orb_params "default ORB feature extraction setting"
[2023-09-01 00:02:20.215] [I] startup SLAM system
[2023-09-01 00:02:20.215] [I] start mapping module
[2023-09-01 00:02:20.215] [I] start global optimization module

and when I stop I get:

[2023-09-01 00:03:19.694] [I] terminate mapping module
[2023-09-01 00:03:19.700] [I] terminate global optimization module
[2023-09-01 00:03:19.700] [I] shutdown SLAM system
[2023-09-01 00:03:19.700] [I] encoding 1 camera(s) to store
[2023-09-01 00:03:19.700] [I] encoding 1 orb_params to store
[2023-09-01 00:03:19.700] [I] encoding 0 keyframes to store
[2023-09-01 00:03:19.700] [I] encoding 0 landmarks to store
[2023-09-01 00:03:19.700] [I] save the MessagePack file of database to /home/ps/ros2_ws/src/stella_vslam_ros/map/map.msg
[2023-09-01 00:03:19.708] [I] clear BoW database
[2023-09-01 00:03:19.708] [I] clear map database

From what I can tell most of this is correct, but the lack of keyframes and landmarks looks like it is not receiving the camera feed.

I have checked with RViz and I can connect to the published topic. Also, should I see a PangolinViewer pop-up, because I don't. Did I miss a set where I should have declared the viewer type?

Thanks
